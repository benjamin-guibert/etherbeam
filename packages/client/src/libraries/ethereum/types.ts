import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string
  price?: BigNumber
  priceHistory: TokenPriceHistory[]
  website?: string
  whitepaper?: string
  github?: string
  linkedin?: string
  facebook?: string
  reddit?: string
  twitter?: string
  telegram?: string
  discord?: string
  actions?: TransactionAction[]
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address
  to?: Address
  value?: BigNumber
  gas: {
    unitPrice?: BigNumber
    limit?: number
    used?: number
    ratio?: number
    price?: BigNumber
  }
  action?: TransactionMethod
  logs?: TransactionMethod[]
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[]

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token
  fromAmount?: BigNumber
  toAddress?: Address | Token
  toAmount?: BigNumber
}
