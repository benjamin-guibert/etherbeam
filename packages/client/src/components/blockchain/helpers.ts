import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from '../../libraries/ethereum/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}
