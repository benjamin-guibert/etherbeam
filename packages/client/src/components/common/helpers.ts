import { faDotCircle, faFile, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/ethereum/types'
import { BigNumber, ethers } from 'ethers'

const ETH_SYMBOL = 'ETH'
const ETH_DECIMALS = 18

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
  withSymbol?: boolean
}

const printAmount = (amount: string, symbol: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals, withSymbol } = options ?? {}
  const decimalsCount = forcedDecimals ?? decimals
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`
  const printedAmount = amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'

  return `${withSymbol ? `${symbol} ` : ''}${printedAmount}`
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_SYMBOL, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { symbol, decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, symbol, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faDotCircle
    default:
      return null
  }
}
