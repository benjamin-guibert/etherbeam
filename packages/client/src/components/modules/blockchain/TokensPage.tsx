import React, { ReactElement, useEffect, useState, useRef } from 'react'
import { useHistory } from 'react-router-dom'
import Table from 'react-bootstrap/Table'
import Loader from 'components/common/Loader'
import { ToastContent } from 'components/Toaster'
import Address from 'components/common/Address'
import Difference from 'components/common/Difference'
import { Token } from 'libraries/ethereum/types'
import { fetchTokens } from 'libraries/ethereum/server'
import { printEtherAmount } from 'components/common/helpers'
import { Col, Row } from 'react-bootstrap'
import Page from 'components/common/Page'

const FETCH_TOKENS_INTERVAL = 30 * 1000

interface TokensPageProps {
  addToast: (toast: ToastContent) => void
}

const TokensPage = ({ addToast }: TokensPageProps): ReactElement => {
  const [tokens, setTokens] = useState<Token[]>([])
  const addToastRef = useRef<(toast: ToastContent) => void>(addToast)
  const history = useHistory()

  useEffect(() => {
    addToastRef.current = addToast
  }, [addToast])

  useEffect(() => {
    const startFetchingTokens = async () => {
      fetchTokens()
        .then((fetchedTokens) => {
          if (!fetchedTokens.length) return

          setTokens(fetchedTokens)
        })
        .catch(() => addToastRef.current({ type: 'error', message: 'Error while fetching tokens.' }))
    }
    const interval = setInterval(startFetchingTokens, FETCH_TOKENS_INTERVAL)

    startFetchingTokens()

    return () => {
      clearInterval(interval)
    }
  }, [])

  return (
    <Page title="Tokens">
      <div>
        {!!tokens.length ? (
          <>
            <Table hover responsive striped variant="dark" bordered size="sm">
              <thead className="text-center">
                <tr>
                  <th className="d-none d-sm-table-cell">Name</th>
                  <th>Price (Ether)</th>
                  <th className="d-none d-md-table-cell">Address</th>
                </tr>
              </thead>
              <tbody>
                {tokens.map((token, index) => {
                  const { hash, name, price, priceHistory } = token

                  return (
                    <tr key={index} className="app-clickable" onClick={() => history.push('/tokens/' + hash)}>
                      <td className="d-none d-sm-table-cell">{name}</td>
                      <td className="text-monospace">
                        <Row>
                          <Col>{price ? printEtherAmount(price) : '--'}</Col>
                        </Row>
                        <Row>
                          {!!priceHistory.length &&
                            priceHistory.map(({ time, ratio }, index) => (
                              <Col xs="auto" key={index} className="pr-0">
                                <Difference time={time} ratio={ratio} />
                              </Col>
                            ))}
                        </Row>
                      </td>
                      <td className="d-none d-md-table-cell text-center">
                        <Address address={token} size="m" short noFlag />
                      </td>
                    </tr>
                  )
                })}
              </tbody>
            </Table>
          </>
        ) : (
          <Loader />
        )}
      </div>
    </Page>
  )
}

export default TokensPage
