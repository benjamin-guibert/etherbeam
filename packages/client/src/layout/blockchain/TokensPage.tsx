import React, { ReactElement, useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import Loader from 'components/Loader'
import { Token } from 'libraries/ethereum/types'
import { fetchTokens } from 'libraries/ethereum/server'
import Page from 'components/Page'
import { printEtherAmount } from 'components/blockchain/helpers'
import Address from 'components/blockchain/Address'
import ClickableTableRow from 'components/ClickableTableRow'
import TokenPriceHistory from 'components/blockchain/TokenPriceHistory'

const FETCH_TOKENS_INTERVAL = 30 * 1000

const LoaderRow = (): ReactElement => {
  return (
    <tr>
      <td>
        <Loader type="dots" />
      </td>
      <td>
        <Loader type="dots" />
      </td>
      <td className="my-d-min-xs">
        <Loader type="dots" />
      </td>
    </tr>
  )
}

const TokensPage = (): ReactElement => {
  const [tokens, setTokens] = useState<Token[]>([])
  const history = useHistory()

  useEffect(() => {
    const startFetchingTokens = async () => {
      fetchTokens().then((fetchedTokens) => {
        if (!fetchedTokens.length) return

        setTokens(fetchedTokens)
      })
    }
    const interval = setInterval(startFetchingTokens, FETCH_TOKENS_INTERVAL)

    startFetchingTokens()

    return () => {
      clearInterval(interval)
    }
  }, [])

  const tokenRows = tokens?.map((token, index) => {
    const { hash, name, price, priceHistory } = token

    return () => (
      <ClickableTableRow key={index} onClick={() => history.push('/tokens/' + hash)}>
        <td>{name}</td>
        <td>
          <span className="my-amount">{price ? printEtherAmount(price) : '--'}</span>
          <TokenPriceHistory priceHistory={priceHistory} className="my-m-0" />
        </td>
        <td className="my-d-min-xs my-text-center">
          <Address address={token} short noFlag />
        </td>
      </ClickableTableRow>
    )
  })

  return (
    <Page title="Tokens">
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>Price (Ether)</th>
            <th className="my-d-min-xs">Address</th>
          </tr>
        </thead>
        <tbody>{!!tokens.length ? tokenRows.map((TokenRow, index) => <TokenRow key={index} />) : <LoaderRow />}</tbody>
      </table>
    </Page>
  )
}

export default TokensPage
