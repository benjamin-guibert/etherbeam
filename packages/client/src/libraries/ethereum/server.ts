import axios from 'axios'
import compact from 'lodash/compact'
import { HttpStatus, getPagination, PaginatedContent } from 'libraries/api'
import { Token, Transaction, USDC_HASH, WETH_HASH } from './types'
import { TokenData, TransactionData } from './serverTypes'
import { parseTokenData, parseTransactionData } from './serverParsers'

const TIMEOUT = 60 * 1000
const TRANSACTIONS_PER_PAGE = 20

const axiosClient = axios.create({
  baseURL: process.env.SERVER_API_HOST,
  headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
  timeout: TIMEOUT,
})

export const fetchTokens = async (): Promise<Token[]> => {
  const { data } = await axiosClient.get<TokenData[]>('tokens')
  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (address: string, withList?: 'actions'): Promise<Token> => {
  const config = {
    validateStatus: (status: number) => [HttpStatus.Ok, HttpStatus.NotFound].includes(status),
  }

  const { data } = await axiosClient.get<TokenData>(`tokens/${address}?list=${withList || ''}`, config)

  return data ? parseTokenData(data) : null
}

export const fetchPendingTransactions = async (page: number): Promise<PaginatedContent<Transaction>> => {
  const response = await axiosClient.get<TransactionData[]>(
    `block_transactions/pending?items=${TRANSACTIONS_PER_PAGE}&page=${page}`
  )

  return {
    items: response.data.map((transaction) => parseTransactionData(transaction)),
    pagination: getPagination(response),
  }
}
