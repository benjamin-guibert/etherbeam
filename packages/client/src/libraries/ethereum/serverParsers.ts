import { BigNumber } from 'ethers'
import {
  TransactionData,
  TokenData,
  AddressData,
  TokenActionData,
  TransactionMethodParameterData,
  TransactionMethodData,
} from './serverTypes'
import {
  Transaction,
  Token,
  HistoryTime,
  Address,
  AddressType,
  TransactionMethodValueType,
  TransactionAction,
  TransactionActionDirection,
  TransactionMethodParameter,
  TransactionMethodParameterType,
  TransactionMethod,
} from './types'
import { createTransactionAddress, getEtherscanUrl } from './ethereum'
import orderBy from 'lodash/orderBy'

const parseBigNumberFromServer = (value: string): BigNumber => (value ? BigNumber.from(value.split('.')[0]) : null)

export const parseAddressData = ({ address_hash, address_type, label }: AddressData): Address => {
  return {
    hash: address_hash,
    type: address_type,
    label,
    url: getEtherscanUrl('address', address_hash),
  }
}

const parseHashData = (hash: string, address?: AddressData | TokenData): Address => {
  if (!address) {
    return {
      hash,
      type: AddressType.Unknown,
      url: getEtherscanUrl('address', hash),
    }
  }

  switch (address.address_type) {
    case AddressType.Token:
      return parseTokenData(address as TokenData)
    default:
      return parseAddressData(address)
  }
}

const parseTransactionMethodValueType = ({
  parameter_type,
  decimal_value,
  addresses,
}: TransactionMethodParameterData): TransactionMethodValueType => {
  switch (parameter_type) {
    case TransactionMethodParameterType.UnsignedInteger:
      return parseBigNumberFromServer(decimal_value)
    case TransactionMethodParameterType.Address:
      return addresses.length > 0 ? parseHashData(addresses[0].address_hash, addresses[0].address) : null
    case TransactionMethodParameterType.Addresses:
      return addresses ? addresses.map((address) => parseHashData(address.address_hash, address.address)) : null
    default:
      return null
  }
}

const parseTransactionMethodParameter = (parameter: TransactionMethodParameterData): TransactionMethodParameter => {
  const { name, parameter_type, raw_type } = parameter

  return {
    name,
    type: parameter_type,
    rawType: raw_type,
    value: parseTransactionMethodValueType(parameter),
  }
}

const parseTransactionMethodData = ({ contract, name, parameters }: TransactionMethodData): TransactionMethod => {
  return {
    contract: parseAddressData(contract),
    name,
    parameters: parameters.map((parameter) => parseTransactionMethodParameter(parameter)),
  }
}

export const parseTransactionData = ({
  transaction_hash,
  status,
  datetime,
  block_number,
  from_address_hash,
  from_address,
  to_address_hash,
  to_address,
  value,
  gas_unit_price,
  gas_limit,
  gas_used,
  gas_ratio,
  gas_fee,
  transaction_method_action,
  transaction_method_logs,
}: TransactionData): Transaction => {
  return {
    blockNumber: block_number,
    dateTime: datetime,
    address: createTransactionAddress(transaction_hash),
    status,
    from: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    to: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    value: parseBigNumberFromServer(value),
    gas: {
      unitPrice: parseBigNumberFromServer(gas_unit_price),
      limit: gas_limit,
      used: gas_used,
      ratio: gas_ratio,
      price: parseBigNumberFromServer(gas_fee),
    },
    action: transaction_method_action ? parseTransactionMethodData(transaction_method_action) : null,
    logs: transaction_method_logs?.map((log) => parseTransactionMethodData(log)),
  }
}

const parseTokenActionData = (
  {
    block_transaction,
    holder_address_hash,
    holder_address,
    index,
    action_type,
    from_address,
    from_address_hash,
    from_amount,
    to_address,
    to_address_hash,
    to_amount,
  }: TokenActionData,
  direction: TransactionActionDirection
): TransactionAction => {
  return {
    transaction: parseTransactionData(block_transaction),
    type: action_type,
    direction,
    index,
    holder: parseHashData(holder_address_hash, holder_address),
    fromAddress: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    fromAmount: parseBigNumberFromServer(from_amount),
    toAddress: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    toAmount: parseBigNumberFromServer(to_amount),
  }
}

export const parseTokenData = (data: TokenData): Token => {
  const {
    address_hash,
    name,
    description,
    symbol,
    decimals,
    price,
    price_1h,
    price_1h_ratio,
    price_1d,
    price_1d_ratio,
    price_1w,
    price_1w_ratio,
    price_1m,
    price_1m_ratio,
    price_1y,
    price_1y_ratio,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    from_transaction_actions,
    to_transaction_actions,
  } = data
  const priceHistory = []
  if (price_1h) {
    priceHistory.push({ time: HistoryTime.Hour, price: parseBigNumberFromServer(price_1h), ratio: price_1h_ratio })
  }
  if (price_1d) {
    priceHistory.push({ time: HistoryTime.Day, price: parseBigNumberFromServer(price_1d), ratio: price_1d_ratio })
  }
  if (price_1w) {
    priceHistory.push({ time: HistoryTime.Week, price: parseBigNumberFromServer(price_1w), ratio: price_1w_ratio })
  }
  if (price_1m) {
    priceHistory.push({ time: HistoryTime.Month, price: parseBigNumberFromServer(price_1m), ratio: price_1m_ratio })
  }
  if (price_1y) {
    priceHistory.push({ time: HistoryTime.Year, price: parseBigNumberFromServer(price_1y), ratio: price_1y_ratio })
  }

  const fromActions = from_transaction_actions
    ? from_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Sell))
    : []
  const toActions = to_transaction_actions
    ? to_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Buy))
    : []
  const actions =
    from_transaction_actions || to_transaction_actions
      ? orderBy([...fromActions, ...toActions], ['transaction.dateTime', 'index'], ['desc', 'desc'])
      : null

  return {
    ...parseAddressData(data),
    type: AddressType.Token,
    url: getEtherscanUrl('token', address_hash),
    name,
    description,
    symbol,
    decimals,
    price: price ? parseBigNumberFromServer(price) : null,
    priceHistory,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    actions,
  }
}
