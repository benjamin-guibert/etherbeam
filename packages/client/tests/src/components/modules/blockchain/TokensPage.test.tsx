import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { mocked } from 'ts-jest/utils'
import { act } from 'react-dom/test-utils'
import { clearMocks, mockTradingViewWidget } from '../../../../helpers'
import { createToken } from '../../../../fixtures/ethereum'

jest.mock('libraries/ethereum/server')
mockTradingViewWidget()

import { fetchTokens } from 'libraries/ethereum/server'
import TokensPage from 'components/modules/blockchain/TokensPage'

const fetchTokensMock = mocked(fetchTokens)
const addToast = jest.fn()

describe('<TokensPage />', () => {
  const tokens = [
    createToken('0x0000000000000000000000000000000000000111'),
    createToken('0x0000000000000000000000000000000000000222'),
    createToken('0x0000000000000000000000000000000000000333'),
  ]

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(async () => {
      clearMocks()

      fetchTokensMock.mockResolvedValue(tokens)

      await act(async () => {
        component = mount(<TokensPage addToast={addToast} />)
        await new Promise(setImmediate)
        component.update()
      })
    })

    it('should fetch tokens', () => expect(fetchTokensMock).toBeCalledTimes(1))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Empty', () => {
    beforeAll(async () => {
      clearMocks()

      fetchTokensMock.mockResolvedValue([])

      await act(async () => {
        component = mount(<TokensPage addToast={addToast} />)
        await new Promise(setImmediate)
        component.update()
      })
    })

    it('should fetch tokens', () => expect(fetchTokensMock).toBeCalledTimes(1))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Error', () => {
    beforeAll(async () => {
      clearMocks()

      fetchTokensMock.mockRejectedValue(new Error('Error'))

      await act(async () => {
        component = mount(<TokensPage addToast={addToast} />)
        await new Promise(setImmediate)
        component.update()
      })
    })

    it('should fetch tokens', () => expect(fetchTokensMock).toBeCalledTimes(1))

    it('should add toast', () => {
      expect(addToast).toBeCalledTimes(1)
      expect(addToast.mock.calls[0][0]).toMatchObject({ type: 'error', message: 'Error while fetching tokens.' })
    })
  })
})
