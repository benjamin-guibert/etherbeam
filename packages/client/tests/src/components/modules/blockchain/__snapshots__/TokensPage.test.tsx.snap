// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokensPage /> Default should match snapshot 1`] = `
<div
  class="container"
>
  <h1>
    Tokens
  </h1>
  <div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-dark table-sm table-striped table-bordered table-hover"
      >
        <thead
          class="text-center"
        >
          <tr>
            <th
              class="d-none d-sm-table-cell"
            >
              Name
            </th>
            <th>
              Price (Ether)
            </th>
            <th
              class="d-none d-md-table-cell"
            >
              Address
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="app-clickable"
          >
            <td
              class="d-none d-sm-table-cell"
            >
              Token
            </td>
            <td
              class="text-monospace"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  NaN
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1H
                    </span>
                    +100 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1D
                    </span>
                    -50 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1W
                    </span>
                    -75 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-secondary"
                  >
                    <span
                      class="mr-2"
                    >
                      1M
                    </span>
                    +0 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1Y
                    </span>
                    +300 %
                  </span>
                </div>
              </div>
            </td>
            <td
              class="d-none d-md-table-cell text-center"
            >
              <a
                class="p-0 app-text-m  btn btn-link"
                href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
                target="_blank"
              >
                <span
                  class="text-monospace"
                >
                  0x0000...0111
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="app-clickable"
          >
            <td
              class="d-none d-sm-table-cell"
            >
              Token
            </td>
            <td
              class="text-monospace"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  NaN
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1H
                    </span>
                    +100 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1D
                    </span>
                    -50 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1W
                    </span>
                    -75 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-secondary"
                  >
                    <span
                      class="mr-2"
                    >
                      1M
                    </span>
                    +0 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1Y
                    </span>
                    +300 %
                  </span>
                </div>
              </div>
            </td>
            <td
              class="d-none d-md-table-cell text-center"
            >
              <a
                class="p-0 app-text-m  btn btn-link"
                href="https://etherscan.io/address/0x0000000000000000000000000000000000000222"
                target="_blank"
              >
                <span
                  class="text-monospace"
                >
                  0x0000...0222
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="app-clickable"
          >
            <td
              class="d-none d-sm-table-cell"
            >
              Token
            </td>
            <td
              class="text-monospace"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  NaN
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1H
                    </span>
                    +100 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1D
                    </span>
                    -50 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-danger"
                  >
                    <span
                      class="mr-2"
                    >
                      1W
                    </span>
                    -75 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-secondary"
                  >
                    <span
                      class="mr-2"
                    >
                      1M
                    </span>
                    +0 %
                  </span>
                </div>
                <div
                  class="pr-0 col-auto"
                >
                  <span
                    class="text-monospace badge badge-success"
                  >
                    <span
                      class="mr-2"
                    >
                      1Y
                    </span>
                    +300 %
                  </span>
                </div>
              </div>
            </td>
            <td
              class="d-none d-md-table-cell text-center"
            >
              <a
                class="p-0 app-text-m  btn btn-link"
                href="https://etherscan.io/address/0x0000000000000000000000000000000000000333"
                target="_blank"
              >
                <span
                  class="text-monospace"
                >
                  0x0000...0333
                </span>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`<TokensPage /> Empty should match snapshot 1`] = `
<div
  class="container"
>
  <h1>
    Tokens
  </h1>
  <div>
    <div
      class="text-center "
    >
      <div
        class="text-center m-1 spinner-border text-secondary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
